*   Multimídia - Compressão II (Exercício 1)

*   Felipe Daniel Dias dos Santos - 11711ECP004

*   Graduação em Engenharia de Computação - Faculdade de Engenharia Elétrica - Universidade Federal de Uberlândia

In [ ]:
#Função para buscar um decimal fracionário, incluído no intervalo (high, low), que possui o valor binário de menor tamanho possível (binário ótimo)
def bestBin(high, low):

  #Variável para armazenamento do valor binário
  code = ""

  #Laço para análise do valor desejado no intervalo correspondente
  while bin2dec(code) < low:

    code = code + "1"

    if bin2dec(code) > high:

      code = code[:len(code) - 1]
      code = code + "0"

  return '0.' + code

In [ ]:
#Função para conversão de um número binário em decimal fracionário
def bin2dec(num):

  #Variável para armazenamento do valor final
  sum = 0.0

  #Variável vazia retorna valor nulo
  if num == "":

    return 0
  
  #Cálculo do decimal fracionário correspondente, caso a variável não seja vazia
  else:
    
    for i in range(len(num)):

      sum = sum + int(num[i]) / (2 ** (i + 1) )

  return sum

In [ ]:
#Função para realizar a soma cumulativa de uma lista de probabilidades "V"
def cumulative(p):

  #Lista para armazenamento da soma acumulada
  acum = []
  acum.append(0.0)

  #Normalização do vetor "p" para o cálculo da soma acumulada
  p = p / np.sum(p)

  #Cálculo da soma acumulada
  for i in range(len(p)):

    acum.append(acum[i] + p[i])

  return np.array(acum)

In [ ]:
#Função que realiza a codificação aritmética de um conjunto de dados "string"
def arithmetic(string):

  #Transformação da string em lista
  string = list(string)

  #Acréscimo do símbolo terminador a string
  string.append('$')

  #Obtenção do alfabeto e da frequência de ocorrência de cada símbolo
  alf, freq = np.unique(string, return_counts = True)

  #Conversão do alfabeto em lista
  alf = list(alf)

  #Obtenção da soma acumulada do vetor de frequências
  p = cumulative(freq)

  #Definição das variáveis para o cálculo da codificação (limite superior, limite inferior e intervalo)
  low = 0.0
  high = 1.0
  r = 1.0

  #Laço que analisa toda a string de dados
  for i in range(len(string)):

    #Obtenção do caractere corrente
    elem = string[i]

    #Obtenção da posição do elemento corrente no alfabeto
    pos = alf.index(elem)
    
    #Obtenção dos limites do caractere corrente no vetor de probabilidades
    range_low = p[pos]
    range_high = p[pos + 1]

    #Atualização das variáveis relevantes (limite superior, limite inferior e intervalo)
    high = low + r * range_high
    low = low + r * range_low
    r = high - low

  #O valor dos limites finais obtidos pelo algoritmo serão enviados para a função "bestBin", que retornará o menor binário contido no intervalo
  return bestBin(high, low)

Utilizando as funções definidas acima, serão construídos 3 strings, onde:

Cada string passará pelo processo de codificação aritmética, resultando em um intervalo único para todo o texto;

Com esse intervalo, vamos obter algum valor contido nele tal que seu correspondente binário possua a menor quantidade possível de bits;

Para medir a eficiência dessa técnica, será calculada a taxa de compressão dos mesmos e será realizada uma comparação com a mesma medida de outras técnicas de compressão.

In [ ]:
#Texto I

#Declaração das bibliotecas necessárias
import numpy as np

#Definição do texto
string = "Johann Sebastian Bach"

print("Texto original:", string)

#Obtenção do tamanho da string
n_string = len(string)

#Codificação do texto de acordo com a codificação aritmética
code = arithmetic(string)

print("\nTexto codificado (decimal):", bin2dec(code[2:]))
print("\nTexto codificado (binário):", code)

#Obtenção do alfabeto do texto
alf, _ = np.unique(list(string), return_counts = True)

#Cálculo da quantidade de bits necessária para representar a informação original sem compressão
bits_o = np.ceil(np.log2(len(alf))) * n_string

#Cálculo da quantidade de bits da representação da informação após compressão (excluindo a parte "0.", pois não é necessária para a decodificação)
bits_f = len(code) - 2

#Cálculo da taxa de compressão
print("\nTaxa de compressão:", bits_o / bits_f)

Texto original: Johann Sebastian Bach

Texto codificado (decimal): 0.22237473103089497

Texto codificado (binário): 0.0011100011101101100011001110010100111100000001111000111

Taxa de compressão: 1.5272727272727273


In [ ]:
#Texto II

#Definição do texto
string = "Wolfgang Amadeus Mozart"

print("Texto original:", string)

#Obtenção do tamanho da string
n_string = len(string)

#Codificação do texto de acordo com a codificação aritmética
code = arithmetic(string)

print("\nTexto codificado (decimal):", bin2dec(code[2:]))
print("\nTexto codificado (binário):", code)

#Obtenção do alfabeto do texto
alf, _ = np.unique(list(string), return_counts = True)

#Cálculo da quantidade de bits necessária para representar a informação original sem compressão
bits_o = np.ceil(np.log2(len(alf))) * n_string

#Cálculo da quantidade de bits da representação da informação após compressão (excluindo a parte "0.", pois não é necessária para a decodificação)
bits_f = len(code) - 2

#Cálculo da taxa de compressão
print("\nTaxa de compressão:", bits_o / bits_f)

Texto original: Wolfgang Amadeus Mozart

Texto codificado (decimal): 0.23994217782629534

Texto codificado (binário): 0.0011110101101100110110011011111010110001111010001001

Taxa de compressão: 2.2115384615384617


In [ ]:
#Texto III

#Definição do texto
string = "Franz Peter Schubert"

print("Texto original:", string)

#Obtenção do tamanho da string
n_string = len(string)

#Codificação do texto de acordo com a codificação aritmética
code = arithmetic(string)

print("\nTexto codificado (decimal):", bin2dec(code[2:]))
print("\nTexto codificado (binário):", code)

#Obtenção do alfabeto do texto
alf, _ = np.unique(list(string), return_counts = True)

#Cálculo da quantidade de bits necessária para representar a informação original sem compressão
bits_o = np.ceil(np.log2(len(alf))) * n_string

#Cálculo da quantidade de bits da representação da informação após compressão (excluindo a parte "0.", pois não é necessária para a decodificação)
bits_f = len(code) - 2

#Cálculo da taxa de compressão
print("\nTaxa de compressão:", bits_o / bits_f)

Texto original: Franz Peter Schubert

Texto codificado (decimal): 0.17676204920572364

Texto codificado (binário): 0.001011010100000001000111000101001000001100110100111001

Taxa de compressão: 1.4814814814814814


Analisando os resultados obtidos, notamos que a taxa de compressão assumiu valores considerados muito bons. Em comparação com o algoritmo de Huffman, obtemos os seguintes valores para a taxa de compressão:

Texto I: Huffman (1.091), Aritmética (1.527); 

Texto II: Huffman (1.211), Aritmética (2.211); 

Texto III: Huffman (1.081), Aritmética (1.481). 

Dessa forma, concluimos que a codificação Aritmética gera uma quantidade de bits menor que a codificação de Huffman. Além disso, de todas as técnicas exploradas (Run Length, Shannon-Fano, Huffman e Aritmética), a que mais se aproxima do valor ótimo (entropia) é a codificação Aritmética. Isso é devido ao fato de que essa codificação gera uma quantidade fracionária de bits por símbolo, enquanto que as demais técnicas geram somente quantidades inteiras de bits, limitando assim sua capacidade de compressão.